pkg/lifecycle/service.go

```diff
@@ -176,6 +187,69 @@ func (s *Service) Start(
 	return nil
 }
 
+// Restart replaces the nodes of a pipeline and starts it again.
+func (s *Service) Restart(ctx context.Context, rp *runnablePipeline) error {
```

Comment:
Thinking aloud... `Restart()` should actually be `Start()`, but with the proper backoff config. It would also be good that a restart starts with a clean slate, which is why I think it's better to just create a new `runningPipeline` then fix the existing one. It's also safer if we add other fields to `runningPipeline` in future.

`Restart()` can call `Start()` and give it the existing `runningPipeline`'s backoff config. Or `Start()` can check if a pipeline is already running, and copy over the backoff config.

Not a must have for this PR, I think it already looks great and we're ready to merge it!

Reply:
I'm giving this one a try and I'm thinking that this is not ideal. In the start method, if the pipeline is running, it will return an error saying we can't start it again. So, if the start method returns that error, the retries will be stopped. We could put it into a degraded state if that was the last attempt

So, in other words, even if when trying to recover it, we return an error when we shouldn't, running should be a good thing. The runnablePipeline also contains the backoff configuration would also require some refactoring so we preserve the backoff altogether.